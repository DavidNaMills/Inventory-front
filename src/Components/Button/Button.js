import React from 'react';
import classes from './button.module.scss';

const allTypes = {
    base: {
        base: classes.button__base,
        shadow: classes.button__base_shadow,
        border: classes.button__base_border,
        skeleton: classes.button__base_skeleton,
    },

    submit: {
        base: classes.button__submit,
        shadow: classes.button__submit_shadow,
        border: classes.button__submit_border,
        skeleton: classes.button__submit_skeleton,
    },

    clear: {
        base: classes.button__clear,
        shadow: classes.button__clear_shadow,
        border: classes.button__clear_border,
        skeleton: classes.button__clear_skeleton,
    },

    danger: {
        base: classes.button__danger,
        shadow: classes.button__danger_shadow,
        border: classes.button__danger_border,
        skeleton: classes.button__danger_skeleton,
    },

    information: {
        base: classes.button__information,
        shadow: classes.button__information_shadow,
        border: classes.button__information_border,
        skeleton: classes.button__information_skeleton,
    },
};


const Button = ({ config = {}, onClick, label = null, skeleton = false, raised = false, full = false, type = 'base', border = false, bold = false, wide = false, isDisabled = false, Icon = null, noStyle=null }) => {
    const fullStyle = [classes.button];
    let which = type;
    const style = {};

    if (isDisabled) {
        fullStyle.push(classes.button__disabled);
    } else if(noStyle) { fullStyle.push(classes.button_noStyle);
    } else {
        if (allTypes[type]) {
            which = type;
            fullStyle.push(allTypes[type].base);
        } else {
            which = 'base';
            fullStyle.push(allTypes['base'].base);
        }

        if (skeleton) { fullStyle.push(allTypes[which].skeleton); }
        if (raised) { fullStyle.push(allTypes[which].shadow); }
        if (border) { fullStyle.push(allTypes[which].border); }
    }
    if (full) { fullStyle.push(classes.button_full); }
    if (bold) { fullStyle.push(classes.button_bold); }
    
    
    if (wide) {
        style.width = '30%';
    }

    return (
        <button
            {...config}
            onClick={onClick}
            style={style}
            className={fullStyle.join(' ')}
            disabled={isDisabled}
        ><p>{label}</p>{Icon && <Icon />}</button>
    )
}

export default React.memo(Button);