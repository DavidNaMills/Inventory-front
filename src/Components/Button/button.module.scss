@import '../../styles/mixins/buttonMixins.module.scss';
@import '../../styles/variables/colors.module.scss';

.button{
    display: flex;
    align-content: center;
    justify-content: center;
    font-size: 16px;
    border: none;
    padding: 10px;
    padding-left: 20px;
    padding-right: 20px;
    border-radius: 2px;

    p{
        padding-top: 5px;
        padding-right: 10px;
    }

    &__disabled{
        background-color: $disabled-back;
        color: $disabled-color;
    }

    &__base{
        @include buttonStyle($base-back, $base-color);
        &:hover{
            @include onHover($base-shadow, $light-text, $base-back);
        }

        &_shadow{@include buttonShadow($base-shadow);}
        &_border{@include buildBorder($base-color);}
        &_skeleton{@include buildSkeleton($base-back, $base-color); }
    };
    
    &__submit{
        @include buttonStyle($submit-back, $submit-color);
        &:hover{
            @include onHover($submit-shadow, $light-text, $submit-back);
        }

        &_shadow{ @include buttonShadow($submit-shadow); }
        &_border{ @include buildBorder($submit-color); }
        &_skeleton{@include buildSkeleton($submit-back, $submit-color); }
    }
    
    &__clear{
        @include buttonStyle($clear-back, $clear-color);
        &:hover{
            @include onHover($clear-shadow, $clear-color, $clear-back);
        }

        &_shadow{ @include buttonShadow($clear-shadow); }
        &_border{ @include buildBorder($clear-color); }
        &_skeleton{@include buildSkeleton($clear-back, $clear-color); }
    }
    
    &__danger{
        @include buttonStyle($danger-back, $danger-color);
        &:hover{
            @include onHover($danger-shadow, $light-text, $danger-back);
        }
        &_shadow{ @include buttonShadow($danger-shadow); }
        &_border{ @include buildBorder($danger-color); }
        &_skeleton{@include buildSkeleton($danger-back, $danger-color); }
    }
    
    &__information{
        @include buttonStyle($information-back, $information-color);
        &:hover{
            @include onHover($information-shadow, $light-text, $information-back);
        }
        &_shadow{ @include buttonShadow($information-shadow); }
        &_border{ @include buildBorder($information-color); }
        &_skeleton{@include buildSkeleton($information-back, $information-color); }
    }

    &_flat{
        border: none;
        box-shadow: none;
        -webkit-box-shadow: none;
        -moz-box-shadow: none;
    }

    &_full{
        width: 100%;
    }

    &_bold{
        font-weight: bold;
    }

    &_noStyle{
        background-color: transparent;
        border: none;
        box-shadow: none;

        &:hover{
            background-color: rgb(212, 212, 212);
            padding-top: 10px;
            padding-bottom: 10px;
            padding-left: 5px;
            padding-right: 10px;
            border-radius: 100px;
        }
    }
}