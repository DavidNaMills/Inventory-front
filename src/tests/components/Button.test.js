import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, mount} from 'enzyme';
import Button from '../../Components/Button/Button';
import MenuBookIcon from '@material-ui/icons/MenuBook';

const mockClick = jest.fn();

afterEach(()=>{
    mockClick.mockClear();
});

const props = {
    onClick: mockClick,
    label: 'test'
};

describe('Button test suite', ()=>{
    describe('behavioural tests', ()=>{
        let wrapper;

        beforeEach(()=>{
            wrapper = shallow(<Button {...props}/>);
        })

        it('matches snapshot', ()=>{
            const comp = renderer.create(<Button {...props}/>)
            const app = comp.toJSON();
            expect(app).toMatchSnapshot();
        });

        it('exists', ()=>{
            expect(wrapper.exists).toBeTruthy();
        });
        
        it('invokes the function when clicked', ()=>{
            const btn = wrapper.find('button')
            expect(btn.length).toBe(1);
            btn.simulate('click');
            expect(mockClick).toHaveBeenCalledTimes(1);
        });

        it('adds the label', ()=>{
            expect(wrapper.find('button').text()).toEqual(props.label);
        });

        it('sets button not disabled as default', ()=>{
            expect(wrapper.find('button').props().disabled).toBeFalsy();
        });
        
        it('removes style from button', ()=>{
            const tw = shallow(<Button {...props} noStyle/>);
            expect(tw.find('.button_noStyle').length).toBe(1);
        });
        
        it('adds the bold style', ()=>{
            const tw = shallow(<Button {...props} bold/>);
            expect(tw.find('.button_bold').length).toBe(1);
        });
        
        it('adds base style as default', ()=>{
            expect(wrapper.find('.button__base').length).toBe(1);
        });

        it('adds an Icon', ()=>{
            const tw = mount(<Button {...props} Icon = {MenuBookIcon}/>)
            expect(tw.find(MenuBookIcon).length).toBe(1);
            expect(tw.find('button').text()).toEqual(props.label);
        });
    });

    describe('styles tests', ()=>{
        const baseCN = (type)=>`button__${type}`;
        const shadow = (type)=>`button__${type}_shadow`;
        const border = (type)=>`button__${type}_border`;
        const skeleton = (type)=>`button__${type}_skeleton`;

        [
            {msg:'sets classNames for BASE type with shadow and border set', props: {...props, type: 'base', raised: true, border: true}},
            {msg:'sets classNames for SUBMIT type with shadow and border set', props: {...props, type: 'submit', raised: true, border: true}},
            {msg:'sets classNames for CLEAR type with shadow and border set', props: {...props, type: 'clear', raised: true, border: true}},
            {msg:'sets classNames for DANGER type with shadow and border set', props: {...props, type: 'danger', raised: true, border: true}},
            {msg:'sets classNames for INFORMATION type with shadow and border set', props: {...props, type: 'information', raised: true, border: true}},
        ].forEach(x=>{
            it(x.msg, ()=>{
                const btnWrap = shallow(<Button {...x.props}/>);
                expect(btnWrap.find(`.${baseCN(x.props.type)}`).length).toBe(1);
                expect(btnWrap.find(`.${border(x.props.type)}`).length).toBe(1);
                expect(btnWrap.find(`.${shadow(x.props.type)}`).length).toBe(1);
            });
        });

        [
            {msg:'sets classNames for BASE type with skeleton and full options set', props: {...props, type: 'base', skeleton: true, full: true}},
            {msg:'sets classNames for SUBMIT type with skeleton and full options set', props: {...props, type: 'submit', skeleton: true, full: true}},
            {msg:'sets classNames for CLEAR type with skeleton and full options set', props: {...props, type: 'clear', skeleton: true, full: true}},
            {msg:'sets classNames for DANGER type with skeleton and full options set', props: {...props, type: 'danger', skeleton: true, full: true}},
            {msg:'sets classNames for INFORMATION type with skeleton and full options set', props: {...props, type: 'information', skeleton: true, full: true}},
        ].forEach(x=>{
            it(x.msg, ()=>{
                const btnWrap = shallow(<Button {...x.props}/>);
                expect(btnWrap.find(`.${baseCN(x.props.type)}`).length).toBe(1);
                expect(btnWrap.find(`.${skeleton(x.props.type)}`).length).toBe(1);
                expect(btnWrap.find('.button_full').length).toBe(1);
            });
        })
    });
});