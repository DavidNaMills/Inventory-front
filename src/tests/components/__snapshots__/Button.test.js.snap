// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test suite behavioural tests matches snapshot 1`] = `
<button
  className="button button__base"
  disabled={false}
  onClick={[MockFunction]}
  style={Object {}}
>
  <p>
    test
  </p>
</button>
`;
